content/blog/2022-05-19-CoC/index.md

```diff
@@ -0,0 +1,298 @@
+  [Alaia Solko-Breslin](https://alaiasolkobreslin.github.io/) is an M.Eng.
+  student interested in programming languages. She also enjoys playing 
+  volleyball and cooking.
+"""
+[[extra.authors]]
+name = "Alaia Solko-Breslin"
+link = "https://alaiasolkobreslin.github.io/"
++++
+
+Code: [here](https://github.com/alaiasolkobreslin/CoC)
+
+## Background
+
+The Calculus of Constructions (CoC) is a type theory created by Thierry Coquand.
+It is a higher-order typed lambda calculus that is the apex of the 
+[lambda cube](https://en.wikipedia.org/wiki/Lambda_cube).
+In CoC it is possible to define functions from terms to terms, types to types, types to terms, 
+and terms to types. Functions being able to go from terms to types is what makes CoC so powerful.
+Types that depend on terms are known as *dependent types*, and most languages
+do not have them. As an example, consider the following vector type in Coq
+where the length of the vector is part of its type.
+
+```coq
+Inductive vec : Type -> nat -> Type :=
+| vnil : forall X : Type, vec X 0
+| vcons : forall X : Type, X -> forall n : nat, vec X n -> vec X (S n).
+
+Definition v0 : vec bool 0 := vnil bool.
+Definition v1 : vec bool 1 := vcons bool true 0 v0.
+Fail Definition v0 : vec bool 1 := vnil bool.
+```
+
+Notice how the last definition fails because using the number 1 in `v0`'s type
+makes the term ill-typed.
+
+CoC is part of the Calculus of Inductive Constructions (CiC), which adds inductive
+types to CoC, and CiC is the basis of Gallina! But for the purposes of this
+project, we only implemented CoC.
+
+
+## Design and Implementation
+
+### Typing rules
+
+Unlike most other languages, CoC eliminates the distinction between types and terms. 
+We describe the CoC syntax below. We use $ M, N, ... $ for terms in general
+and $ x, y, z $ for variables.
+<p align="center">
+    $ M ::= x \; | \; (\lambda x:M) N \; | \; (M N) \; | \; [x:M] N \; | \; * $
+</p>
+
+$ [x:M] N $ is called a product, and $ * $ is the universe of all types, but is not a
+type itself. For an explanation of why, see the [Bonus section](#bonus).
+*Contexts* (denoted by $ \Gamma, \Delta $) are products over $*$ and have the
+form $[x_1:M_1] ... [x_n:M_n] *$. The inference system defines two judgements:
+
+1. $ \Gamma \vdash \Delta $ means $ \Delta $ is a valid context in the valid
+   context $ \Gamma $.
+2. $ \Gamma \vdash M : P $ means $ M $ is a well-typed term of type $ P $ in
+   the valid context $ \Gamma $
+
+In the inference system, a type $ P $ will either be a context or have the 
+property that $ \Gamma \vdash P : * $. Also note that $ [N/x]Q $ is the notation
+for substituting $N$ for $x$ in $Q$. The inference rules are below.
+
+<p align="center">
+<img src="fig1.png" alt="alt_text" title="image_tooltip" style="zoom:25%;" />
+</p>
+
+### Design
+
+When I started this project, I was writing my implementation based off of
+some notes written by Professor Clarkson, and this was the syntax of CoC that he used.
```

Comment:
You could consider linking to [Clarkson's webpage](https://www.cs.cornell.edu/~clarkson/) here.